### ISSUE TITLE
Specify `is_training` parameter for batch normalization.

### ISSUE THREAD

**Issue by KosukeArase** [tensorflow/hub#147]
Hi, specifying training/inference mode is sometimes necessary especially for batch normalization.
However, it seems that `hub.Module()` does not have `training` or `is_training` arg.
How can I set the training/inference mode for `hub.Module()`?

**Comment by arnoegw** [comment]
Hi, thanks for reaching out.

Short answer: We recommend to use `hub.Module(..., tags=...)` to make this distinction

Long answer: Yes, techniques like dropout and batch normalization require different TensorFlow graphs for training and inference. A TensorFlow Hub Module can store several graph versions, each identified by a set of tags (i.e., strings). By convention, the modules on https://tfhub.dev/ use the empty set of tags for inference, and `tags={"train"}` for training/fine-tuning *if* that requires a different graph version.

When using a module, a graph version is selected by passing a set of strings as the optional `tags=...` argument of `hub.Module(...)`. By default, this is the empty set of tags.

Please note that the `tags=...` parameter for the graph version is separate from the `trainable=...` parameter, much like `tf.layers.batch_normalization()` has separate `training=...` and `trainable=...` parameters.

When calling `hub.create_module_spec` to create a module from a module_fn, the optional argument `tags_and_args=...` can be used to specify the module's tag sets, and how to call the module_fn for each of them. A typical example looks like

```
def your_module_fn(training):
   net = ....  # Build up your module contents.
   net = tf.layers.dropout(net, rate=..., training=training)
   ... # More code here.

module_spec = hub.create_module_spec(
    your_module_fn,
    tags_and_args=[(set([]), dict(training=False)),
                   (set(["train"]), dict(training=True))])
```

Exporting a Module exports all the graph versions described in its spec, not just the one it has been instantiated with.

Hope that helps.

**Comment by KosukeArase** [comment]
@arnoegw Thank you for the clear explanation!
So you mean we need to add `tags={"train"}` only for training and `tags` is unnecessary for inference, right?
When I do so, the accuracy was significantly decayed (nearly chance rate). When I add `tags={"train"}` for testing, the score improved but not as high as the same model from slim.
Do you have any idea why my model doesn't work without `tags` for inference? Thank you.

**Comment by arnoegw** [comment]
Hi @KosukeArase 

Yes, `tags={'train'}` is for training and `tags` unset is for inference. For the image modules on tfhub.dev, that distinction is mostly about operating batch normalization for training or inference. This means `tags={'train'}` will use per-batch statistics for normalization while providing UPDATE_OPS to aggregate dataset statistics behind the scenes. Unset `tags` use those dataset statistics aggregrated at training time.

When training, it is important to execute the UPDATE_OPS with each training step, like `tf.contrib.training.create_train_op()` does, in order to keep the dataset statistics in updated with the changing model. If that happens, the effect at evaluation time can be as you describe: the outdated dataset statistics perform really bad, the ad-hoc statistics from the current batch perform so-so, but only the correct dataset statistics would achieve the expected quality.
